### PR TITLE
NIFI-9229 Flow upgrade not possible if a Output Port changes to a funnel

### PR DESCRIPTION
**Pull request by timeabarna**
https://issues.apache.org/jira/browse/NIFI-9229

#### Description of PR

Fixing bug NIFI-9229

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [x] Is there a JIRA ticket associated with this PR? Is it referenced 
     in the commit message?

- [x] Does your PR title start with **NIFI-XXXX** where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.

- [x] Has your PR been rebased against the latest commit within the target branch (typically `main`)?

- [x] Is your initial contribution a single, squashed commit? _Additional commits in response to PR reviewer feedback should be made on this branch and pushed to allow change tracking. Do not `squash` or use `--force` when pushing to allow for clean monitoring of changes._

### For code changes:
- [ ] Have you ensured that the full suite of tests is executed via `mvn -Pcontrib-check clean install` at the root `nifi` folder?
- [x] Have you written or updated unit tests to verify your changes?
- [x] Have you verified that the full build is successful on JDK 8?
- [ ] Have you verified that the full build is successful on JDK 11?
- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)? 
- [ ] If applicable, have you updated the `LICENSE` file, including the main `LICENSE` file under `nifi-assembly`?
- [ ] If applicable, have you updated the `NOTICE` file, including the main `NOTICE` file found under `nifi-assembly`?
- [ ] If adding new Properties, have you added `.displayName` in addition to .name (programmatic access) for each of the new properties?

### For documentation related changes:
- [ ] Have you ensured that format looks appropriate for the output in which it is rendered?

### Note:
Please ensure that once the PR is submitted, you check GitHub Actions CI for build issues and submit an update to your PR as soon as possible.

### CODE REVIEW

File: nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-components/src/main/java/org/apache/nifi/groups/StandardProcessGroup.java [hunk 1]
```diff
@@ -4293,16 +4294,30 @@ private void updateProcessGroup(final ProcessGroup group, final VersionedProcess
...
+        while (inputPortsRemovedIterator.hasNext()) {
+            final String removedVersionedId = inputPortsRemovedIterator.next();
             final Port port = inputPortsByVersionedId.get(removedVersionedId);
             LOG.info("Removing {} from {}", port, group);
-            group.removeInputPort(port);
+            try {
+                group.removeInputPort(port);
+                inputPortsRemovedIterator.remove();
+            } catch (IllegalStateException e) {
+                LOG.info("Removing {} from {} not possible at the moment, connection needs to be updated first.", port, group);
```
Comment (adenes) [inline nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-components/src/main/java/org/apache/nifi/groups/StandardProcessGroup.java:4306]: Even though this is an info log message at first sight my impression was that I as a user need to "update the connections". Please update this to something like "...not possible at the moment, will try again after updating the connections."

File: nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-components/src/main/java/org/apache/nifi/groups/StandardProcessGroup.java [hunk 2]
```diff
@@ -4293,16 +4294,30 @@ private void updateProcessGroup(final ProcessGroup group, final VersionedProcess
...
+        while (outputPortsRemovedIterator.hasNext()) {
+            final String removedVersionedId = outputPortsRemovedIterator.next();
             final Port port = outputPortsByVersionedId.get(removedVersionedId);
             LOG.info("Removing {} from {}", port, group);
-            group.removeOutputPort(port);
+            try {
+                group.removeOutputPort(port);
+                outputPortsRemovedIterator.remove();
+            } catch (IllegalStateException e) {
+                LOG.info("Removing {} from {} not possible at the moment, connection needs to be updated first.", port, group);
```
Comment (adenes) [inline nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-components/src/main/java/org/apache/nifi/groups/StandardProcessGroup.java:4319]: same as above

File: nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-components/src/main/java/org/apache/nifi/groups/StandardProcessGroup.java [hunk 3]
```diff
@@ -4343,6 +4358,18 @@ private void updateProcessGroup(final ProcessGroup group, final VersionedProcess
             group.removeFunnel(funnel);
         }
 
+        for (final String removedVersionedId : inputPortsRemoved) {
```
Comment (adenes) [inline nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-components/src/main/java/org/apache/nifi/groups/StandardProcessGroup.java:4361]: No strong opinion on this, but maybe a log message or a comment would be useful here: "removing the remaining input and output ports".

File: nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-components/src/main/java/org/apache/nifi/groups/StandardProcessGroup.java [hunk 4]
```diff
@@ -4730,7 +4757,8 @@ private Connectable getConnectable(final ProcessGroup group, final ConnectableCo
                 // if the flow doesn't contain the properly mapped group id, we need to search all child groups.
                 return group.getProcessGroups().stream()
                     .flatMap(gr -> gr.getInputPorts().stream())
-                    .filter(component -> id.equals(component.getVersionedComponentId().orElse(
+                    .filter(component
```
Comment (adenes) [inline nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-components/src/main/java/org/apache/nifi/groups/StandardProcessGroup.java:4760]: minor: unnecessary formatting change, please revert

**Review by adenes — CHANGES_REQUESTED** [review]
Thanks @timeabarna for the fix. Overall the change looks good to me and works as expected. I only had a few minor comments.

**Comment by timeabarna** [comment]
Thanks @adenes for your help, I've modified the code based on your recommendations.

**Review by adenes — APPROVED** [review]
Thanks @timeabarna for the follow-up commit, the fix looks good to me. Merging to main.

**Pull request merged**